src/proc/syncer-processing-block.cpp

```diff
@@ -10,63 +10,116 @@
+            return;
+        }
 
-            LOG_DEBUG(ss.str());
-            env.matches.enqueue(std::move(f));
-        });
-
-        auto f = [&](frame_holder frame, synthetic_source_interface* source)
         {
-            // if the syncer is disabled passthrough the frame
-            bool enabled = false;
-            size_t n_opts = 0;
-            for( auto& wopt : _enable_opts )
+            std::lock_guard< std::mutex > lock( _mutex );
+
+            if( !_matcher )
             {
-                auto opt = wopt.lock();
-                if( opt )
-                {
-                    ++n_opts;
-                    if( opt->is_true() )
-                    {
-                        enabled = true;
-                        break;
-                    }
-                }
+                if(!create_matcher( frame, log ))
+                    get_source().frame_ready(std::move(frame));
             }
-            if( n_opts  &&  ! enabled )
+            else
             {
-                get_source().frame_ready( std::move( frame ) );
-                return;
+                auto env = syncronization_environment{ source, matches, log };
+                _matcher->dispatch(std::move(frame), env);
             }
+                
+        }
 
-            single_consumer_frame_queue<frame_holder> matches;
+        frame_holder f;
+        {
+            std::unique_lock< std::mutex > lock(_callback_mutex, std::try_to_lock);
+            if (!lock.owns_lock())
+                return;
 
+            while( matches.try_dequeue( &f ) )
             {
-                std::lock_guard<std::mutex> lock(_mutex);
-                _matcher->dispatch(std::move(frame), { source, matches });
+                get_source().frame_ready( std::move( f ) );
             }
+        }
+    };
 
-            frame_holder f;
-            while (matches.try_dequeue(&f))
-                get_source().frame_ready(std::move(f));
+    set_processing_callback( std::shared_ptr< rs2_frame_processor_callback >(
+        new internal_frame_processor_callback< decltype( f ) >( f ) ) );
+}
 
-        };
+bool syncer_process_unit::create_matcher( const frame_holder & frame, bool log )
+{
+    try
+    {
+        auto sensor = frame.frame->get_sensor().get();
+        if (!sensor)
+        {
+            LOG_DEBUG("Sensor is not exist any more cannot create matcher the frame will passthrough ");
+            return false;
+        }
+
+        const device_interface * dev = nullptr;
+        dev = sensor->get_device().shared_from_this().get();
+        if (dev)
+        {
+            _matcher = dev->create_matcher(frame);
+
+            _matcher->set_callback([this](frame_holder f, const syncronization_environment& env) {
+                if (env.log)
+                {
+                    std::stringstream ss;
+                    ss << "SYNCED: ";
+                    auto composite = dynamic_cast<composite_frame *>(f.frame);
```

Comment:
Do we know that's it's always a composite frame??

Reply:
yes the matcher output  should always be composite.

Reply:
I fixed it because , due to the change of the creation of matcher , on playback we will get singles frames.